addons/controllers/calico/calicoconfig_controller.go

```diff
@@ -6,43 +6,269 @@ package controllers
 
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
 	cniv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cni/v1alpha1"
 )
 
-// CalicoConfigReconciler reconciles a CalicoConfig object
+// CalicoConfigReconciler reconciles CalicoConfig resource
 type CalicoConfigReconciler struct {
 	client.Client
 	Log    logr.Logger
 	Scheme *runtime.Scheme
+	Ctx    context.Context
 }
 
 //+kubebuilder:rbac:groups=cni.tanzu.vmware.com,resources=calicoconfigs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=cni.tanzu.vmware.com,resources=calicoconfigs/status,verbs=get;update;patch
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the CalicoConfig object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 func (r *CalicoConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("calicoconfig", req.NamespacedName)
+	log := r.Log.WithValues("CalicoConfig", req.NamespacedName)
 
-	// TODO(user): your logic here
+	log.Info("Start reconciliation")
+
+	// fetch CalicoConfig resource, ignore not-found errors
+	calicoConfig := &cniv1alpha1.CalicoConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, calicoConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("CalicoConfig resource '%v' not found", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// config resources are expected to have the same name as the cluster. However, we ideally try to read the cluster name from the owner reference of the addon config object
+	clusterName := req.NamespacedName.Name
+	for _, ownerRef := range calicoConfig.GetOwnerReferences() {
+		if ownerRef.APIVersion == clusterapiv1beta1.GroupVersion.String() && ownerRef.Kind == "Cluster" {
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+
+	// verify that the cluster with the same name is present
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: clusterName}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("Cluster resource '%s' not found. Cleaning up orphaned resource CalicoConfig '%s'", clusterName, calicoConfig.Name))
+			if err := r.Client.Delete(ctx, calicoConfig); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Deleted CalicoConfig '%s'", calicoConfig.Name))
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// reconcile CalicoConfig resource
+	if retResult, err := r.ReconcileCalicoConfig(cluster, calicoConfig.DeepCopy()); err != nil {
```

Comment:
Why do we need `Deepcopy` here?

Reply:
Per our discussion, moved  deepcopy up after initially reading it from api server